Fix AbortController memory leak with event

tbvjaos510 opened the pull request:
#### 1f6b6bea2079964f2f70dd12df7f4b7fbc335aa2
<pre>
Fix AbortController memory leak with event

<a href="https://bugs.webkit.org/show_bug.cgi?id=298499">https://bugs.webkit.org/show_bug.cgi?id=298499</a>

Reviewed by Ryosuke Niwa.

AbortSignals with event listeners were never garbage collected due to incorrect reachability check in JSAbortSignalOwner::isReachableFromOpaqueRoots().
The method unconditionally returned true for signals with event listeners, even when they had no other references.

Changed the logic to check containsWebCoreOpaqueRoot() before marking the signal as reachable, allowing proper garbage collection.

* LayoutTests/fast/dom/gc-abort-controller-expected.txt: Added.
* LayoutTests/fast/dom/gc-abort-controller.html: Added.
* Source/WebCore/bindings/js/JSAbortSignalCustom.cpp:
(WebCore::JSAbortSignalOwner::isReachableFromOpaqueRoots):

Canonical link: <a href="https://commits.webkit.org/299758@main">https://commits.webkit.org/299758@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7142f1f82c6cc13e24ef04ca8d877459f0728848

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/120100 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/39794 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/30445 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/126444 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/72173 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/121976 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/40490 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/48371 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/91209 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/60514 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/123052 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/32341 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/107687 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/71760 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/31377 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/25792 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/70072 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/101820 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/25979 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/129353 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/47020 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/35667 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/99832 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/47386 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/103874 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/99675 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25306 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/45126 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/23149 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/43651 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/46882 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/52588 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/46350 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/49697 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/48034 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Ahmad-S792 commented on LayoutTests/fast/dom/gc-abort-controller.html:
```diff
@@ -0,0 +1,62 @@
+<head>
+    <meta charset="utf-8">
+    <title>Test AbortController memory leak with event listeners</title>
+    <script src="../../resources/js-test-pre.js"></script>
```
Please use 'js-test.js' instead of 'pre' and 'post.js' scripts.

tbvjaos510 replied:
Updated test to use js-test.js script as requested. Thanks

cdumez commented on LayoutTests/fast/dom/gc-abort-controller.html:
```diff
@@ -0,0 +1,61 @@
+</head>
+<body>
+<script>
+jsTestIsAsync = true;
```
Please add a `description("Test AbortController memory leak with event listeners");` before this line and rebaseline the test.

cdumez commented on LayoutTests/fast/dom/gc-abort-controller.html:
```diff
@@ -0,0 +1,61 @@
+            }
+        }
+        
+        if (aliveCount === 0) {
```
Our GC is conservative. As a result, we never write GC tests that assume all objects get collected. We write tests that *some* objects get collected.

tbvjaos510 replied:
Updated the GC collection condition from 0 to 500 or fewer. Thanks!

cdumez commented on LayoutTests/fast/dom/gc-abort-controller.html:
```diff
@@ -0,0 +1,61 @@
+            clearInterval(handle);
+            testPassed("All 500 AbortSignals were garbage collected.");
+            finishJSTest();
+        } else if (cnt >= 500) {
```
This check is also risky. No guarantee the first gc collects any objects. We usually set use a setTimeout for the failure case.

tbvjaos510 replied:
Adjusted GC and test check timing with setTimeout. I had been using setInterval to retry GC, but I agree it was unintuitive. Thanks!

rniwa replied:
No, Chris's point isn't so much that we need to trigger multiple GC's. It's that GC may not collect some objects because it's conservative. We need to allocate multiple `AbortSingal` objects and check that at least one of them is collected instead of allocating just one object and asserting that it gets GC'ed.

tbvjaos510 replied:
Applied your suggestion to update the test.
Instead of repeatedly triggering GC, I now create 20 AbortSignal objects and check if some of them get collected after a single GC run.
This accounts for the conservative nature of the GC where not all objects may be collected in one pass.

cdumez commented on LayoutTests/fast/dom/gc-abort-controller.html:
```diff
@@ -0,0 +1,58 @@
+    
+    controllers = null;
+
+    setTimeout(() => {
```
This is still likely to be flaky sorry. You can always use an existing gc test as example but essentially, you'll need a setInterval() to keep trying to `gc()` & check until some objects are collected (or a timeout is reached, in which case the test failed).

tbvjaos510 replied:
Updated the test to use the setInterval() pattern as recommended. I refactored it to follow the exact same pattern as `reporting-observer-lifetime.html`

rniwa commented on LayoutTests/fast/dom/gc-abort-controller.html:
```diff
@@ -0,0 +1,51 @@
+        gc();
+        cnt++;
+        
+        if (!weakSignal.deref() && !weakCallback.deref()) {
```
Since we have a stack scanning conservative GC, it's not guaranteed that the single object you created will be GC'ed. Instead, create something like 20 objects and check that at least one of them is GC'ed.

rniwa commented on Source/WebCore/bindings/js/JSAbortSignalCustom.cpp:
```diff
@@ -55,9 +55,10 @@ bool JSAbortSignalOwner::isReachableFromOpaqueRoots(JSC::Handle<JSC::Unknown> ha
-            if (reason) [[unlikely]]
-                *reason = "Has Abort Event Listener"_s;
-            return true;
+            bool isReachable = containsWebCoreOpaqueRoot(visitor, abortSignal);
```
Where do we ever add `abortSignal` as an opaque root?

tbvjaos510 replied:
The existing code already had this containsWebCoreOpaqueRoot check for the no-callback case, so I applied the same logic when event listeners are present.
If I understand correctly, the opaque root gets added through AbortController::opaqueRoot() returning the signal
Let me know if I'm missing something here.

rniwa replied:
oh, I see. The IDL file has `GenerateAddOpaqueRoot` specified.

rniwa commented on LayoutTests/fast/dom/gc-abort-controller.html:
```diff
@@ -0,0 +1,57 @@
+        setTimeout(() => {
+            let aliveCount = 0;
+            for (let weakRef of weakRefs) {
+                if (weakRef.deref()) {
```
No curly braces around a singe line statement.

rniwa commented on LayoutTests/fast/dom/gc-abort-controller.html:
```diff
@@ -0,0 +1,57 @@
+            }
+            
+            if (aliveCount < 20) {
+                testPassed(`Some AbortSignals were garbage collected.`);
```
Ditto.

rniwa commented on LayoutTests/fast/dom/gc-abort-controller.html:
```diff
@@ -0,0 +1,57 @@
+            if (aliveCount < 20) {
+                testPassed(`Some AbortSignals were garbage collected.`);
+            } else {
+                testFailed(`No AbortSignals were collected.`);
```
Ditto.

rniwa commented on LayoutTests/fast/dom/gc-abort-controller.html:
```diff
@@ -0,0 +1,57 @@
+            for (let weakRef of weakRefs) {
+                if (weakRef.deref())
+                    aliveCount++;
+                
```
Remove this blank line?

rniwa commented on LayoutTests/fast/dom/gc-abort-controller.html:
```diff
@@ -0,0 +1,57 @@
+<!DOCTYPE html>
+<html>
+<head>
+    <meta charset="utf-8">
```
No need to specify utf-8 since this file only contains ASCII.

rniwa commented on LayoutTests/fast/dom/gc-abort-controller.html:
```diff
@@ -0,0 +1,57 @@
+<html>
+<head>
+    <meta charset="utf-8">
+    <title>Test AbortController memory leak with event listeners</title>
```
We typically don't specify title in a test.

tbvjaos510 replied:
Thanks! Applied all the requested changes.

rniwa commented on LayoutTests/fast/dom/gc-abort-controller.html:
```diff
@@ -0,0 +1,54 @@
+    setTimeout(() => {
+        gc();
+        
+        setTimeout(() => {
```
No need to wait for 20ms. Just run this code immediately after `gc()`. `gc()` is sync.

rniwa commented on LayoutTests/fast/dom/gc-abort-controller.html:
```diff
@@ -0,0 +1,54 @@
+    }, 0)
+}
+
+onload = () => runTest();
```
`onload = runTest;` will suffice.

tbvjaos510 committed:
Fix AbortController memory leak with event

https://bugs.webkit.org/show_bug.cgi?id=298499

Reviewed by Ryosuke Niwa.

AbortSignals with event listeners were never garbage collected due to incorrect reachability check in JSAbortSignalOwner::isReachableFromOpaqueRoots().
The method unconditionally returned true for signals with event listeners, even when they had no other references.

Changed the logic to check containsWebCoreOpaqueRoot() before marking the signal as reachable, allowing proper garbage collection.

* LayoutTests/fast/dom/gc-abort-controller-expected.txt: Added.
* LayoutTests/fast/dom/gc-abort-controller.html: Added.
* Source/WebCore/bindings/js/JSAbortSignalCustom.cpp:
(WebCore::JSAbortSignalOwner::isReachableFromOpaqueRoots):

Canonical link: https://commits.webkit.org/299758@main
